[turner.lee] 지하철 서비스 미션 리뷰 부탁드립니다.

sumniy opened the pull request:
안녕하세요 리뷰어님. 터너입니다.

step3 까지 진행 후 리뷰 요청 드립니다. 잘 부탁드립니다.

sumniy committed:
Docs: 기능 목록 추가

README.md에 step 1 관련 기능 목록 추가

sumniy committed:
Fix: 기본 뼈대 코드 버그 수정

기본 코드 그대로 테스트 실행 시 애플리케이션이 작동하지 않음.
문제되는 부분 수정.

sumniy committed:
Feat: Step 1 경로 조회 기능 구현

source역과 target역의 id를 통해서 두 역간의
최단 경로를 구해주는 API 구현.

두 역이 연결되지 않아서 경로를 찾을 수 없으면 400을 반환함.

kakao-tars-case committed:
test : Lines와 Path에 대한 테스트 추가

kakao-tars-case committed:
feat : email로 사용자를 찾는 기능 추가

kakao-tars-case committed:
feat : AuthenticationPrincipalArgumentResolver 구현

kakao-tars-case committed:
feat : 토큰을 이용한 사용자 인증기능 구현

kakao-tars-case committed:
feat : 즐겨찾기 추가 기능 구현

kakao-tars-case committed:
feat : 즐겨찾기 추가 기능 구현

kakao-tars-case committed:
feat : 즐겨찾기 조회 기능 구현

kakao-tars-case committed:
feat : 즐겨찾기 삭제 기능 구현

kakao-tars-case committed:
feat : 경로 조회 시 Edge에 요금 포함

kakao-tars-case committed:
feat : 요금 계산 기능

sumniy committed:
Fix: 요금 계산식 오류 수정

할인 요금을 계산하는 식에 오류가 있어 수정.
350원을 공제 후 20%(청소년), 50%(어린이) 할인을 하는게 아니라,
350원 공제 후 20%, 50% 한 금액을 전체 요금에서 빼도록 수정.

sumniy committed:
Style: 코드 리포맷

sumniy committed:
Refactor: path 코드 리팩토링 및 상수값 포장

중복되는 코드를 메서드로 분리.
요금계산에 필요한 수치들을 의미있는 상수들로 포장

sumniy committed:
Refactor: Favorite domain 구조 변경

Favorite 도메인이 source, target에 대한 id가 아닌
Station 객체로 가지고 있게하여 즐겨찾기 목록 조회 시
반복적인 Station find 쿼리를 없앰.

sumniy committed:
Style: 코드 포매팅

sumniy committed:
Fix: 로그인 관련 버그 수정

테스트는 통과하지만 실제 웹 상에서 로그인 시도에서
jwtTokenProvide에서 NoClassDefFoundError가 발생.
이를 해결하기 위해
runtimeOnly 'javax.xml.bind:jaxb-api:2.3.1'
의존성을 추가하고 재빌드함.

htrucci commented on build.gradle:
```diff
@@ -33,6 +33,7 @@ dependencies {
     testImplementation 'org.springframework.boot:spring-boot-starter-test'
 
     runtimeOnly 'com.h2database:h2'
+    runtimeOnly 'javax.xml.bind:jaxb-api:2.3.1'
```
혹시 이건 어떤 내용으로 추가된 걸까요?

sumniy replied:
백엔드 테스트에서는 로그인 관련 기능이 정상적으로 작동하는데, 프론트에서 요청할 시에 에러가 발생해서 해결하기 위해 추가했습니다.

htrucci commented on src/main/java/subway/AuthenticationPrincipalConfig.java:
```diff
@@ -1,10 +1,10 @@
+import subway.auth.ui.AuthenticationPrincipalArgumentResolver;
 
 import java.util.List;
 
```
java.subway 패키지내에 Config, Controller, ExceptionHandler 등이 위치하고 있는데요.
디폴트 코드라서 그냥 두셨을수도 있지만, Application만 놔두고 나머지들은 패키지 내로 위치해주면 좋을 것 같습니다 😄

sumniy replied:
넵 알겠습니다!

htrucci commented on src/main/java/subway/SubwayAdvice.java:
```diff
@@ -0,0 +1,18 @@
+import subway.auth.exception.AuthException;
+
+@RestControllerAdvice
+public class SubwayAdvice {
+
+    @ExceptionHandler(AuthException.class)
+    public ResponseEntity<Void> handleAuthException(Exception e) {
+        e.printStackTrace();
+
+        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
+    }
+}
```
ExceptionHandler로 잘 구현해주셨네요 👍 
다만 Advice는 너무 포괄적인 용어라서 ExceptionHandler 등으로 명명해주시는 것이 좋지 않을까요?

htrucci replied:
ResponseEntity<Void> 타입의 Entity를 리턴하지 않고 아래 어노테이션을 붙여 동작하게 할 수 있습니다 😄 
@ResponseStatus(HttpStatus.UNAUTHORIZED)

sumniy replied:
말씀해주신대로 반영했습니다. 감사합니다!

htrucci commented on src/main/java/subway/auth/application/AuthService.java:
```diff
@@ -1,4 +1,39 @@
+@Service
 public class AuthService {
+
+    private final MemberService memberService;
+    private final JwtTokenProvider jwtTokenProvider;
+
+    public AuthService(MemberService memberService, JwtTokenProvider jwtTokenProvider) {
+        this.memberService = memberService;
+        this.jwtTokenProvider = jwtTokenProvider;
+    }
```
생성자 주입시 주입되는 객체가 두개 이상인 경우 생성자에 @Autowired를 붙이도록 스프링 재단에서 가이드하고 있습니다.
참고 부탁드립니다~!

sumniy replied:
아하 감사합니다.

htrucci commented on src/main/java/subway/auth/application/AuthService.java:
```diff
@@ -1,4 +1,39 @@
+        this.jwtTokenProvider = jwtTokenProvider;
+    }
+
+    public String login(String email, String password) {
+        Member member = memberService.findMemberByEmail(email);
+        if (!member.isSamePassword(password)) {
+            throw new InvalidPasswordException();
+        }
+        return jwtTokenProvider.createToken(email);
+    }
```
email만으로 db에서 password값을 가져온 후 비교하게 되면 뭐,, 가져온 password hash가 로그에 찍힐수도있고,, 
무심코 보안이슈가 생길 수 있으니, 쿼리에서 아예 패스워드까지 비교하는 방법은 어떨까요?

htrucci replied:
요즘은 계정통합, 소셜계정 등으로 이메일이 중복되는 경우도 있어서 (물론 지하철 서비스는 유일하지만 ^^;;)
email말고 PK인 member.id 를 이용하여 토큰을 생성하는건 어떨까요?

sumniy replied:
넵 두 가지 모두 반영해보겠습니다.

htrucci commented on src/main/java/subway/auth/application/AuthService.java:
```diff
@@ -1,4 +1,39 @@
+
+        return LoginMember
+                .of(memberService.findMemberByEmail(jwtTokenProvider.getPayload(accessToken)));
+    }
```
👍

htrucci commented on src/main/java/subway/auth/exception/NoSuchEmailException.java:
```diff
@@ -0,0 +1,10 @@
+package subway.auth.exception;
+
+public class NoSuchEmailException extends AuthException {
+
+    private static final String NO_SUCH_EMAIL_ERROR = "해당하는 이메일이 존재하지 않습니다.";
+
+    public NoSuchEmailException() {
+        super(NO_SUCH_EMAIL_ERROR);
+    }
+}
```
이메일이 없는 케이스, 패스워드가 일치하지 않는 케이스 Exception나누어 상세하게 잘 구현해주셨네요. 👍 
AuthException을 상속받고 Handler로 잘 잡아서 처리해주셨다고 생각되는 부분인데
해당 메시지를 유저에게 그대로 노출해버리면 BruteForce 공격이 가능하기때문에 내부에서는 상세하게 Exception을 내더라도
유저에게 나가는 메시지는 별도로 처리되도록 유의해야합니다.  🤔 
요즘은 "아이디 및 패스워드가 일치하지 않습니다" 등 최소한으로 알려주는 추세입니다 😄

sumniy replied:
넵 참고하겠습니다. 감사합니다.

htrucci commented on src/main/java/subway/auth/ui/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -22,8 +30,21 @@ public boolean supportsParameter(MethodParameter parameter) {
+            return authService.findMember(accessToken);
+        } catch (InvalidTokenException e) {
+            validateLoginMemberNecessary(parameter.getParameterAnnotation(AuthenticationPrincipal.class));
+            return LoginMember.of(new Member(null, null, null, 20));
```
바로 throw new InvalidTokenException(); 해야 될 상황으로 보이는데요.
new Member(null, null, null, 20) 객체를 만들어 리턴하는게 어떤 의미가 있을까요?

>> 20세 셋팅이 필요한 것으로 요건 재확인했습니다.ㅎㅎ

htrucci commented on src/main/java/subway/favorite/application/FavoriteService.java:
```diff
@@ -0,0 +1,38 @@
+        this.stationDao = stationDao;
+    }
+
+    public Favorite createFavorite(Long memberId, Long sourceStationId, Long targetStationId) {
+        Favorite result = favoriteDao.insert(memberId, sourceStationId, targetStationId);
+
+        return new Favorite(
+                result.getId(),
+                result.getMemberId(),
+                stationDao.findById(result.getSourceStationId()),
+                stationDao.findById(result.getTargetStationId()));
+    }
```
이왕 stationDao.findById()를 한다면 insert하기전에 먼저 하면 요청된 stationId가 실제 존재하는지 유효성 검증도 같이 할 수 있겠네요. 👀

sumniy replied:
그렇네요! 감사합니다.

htrucci commented on src/main/java/subway/favorite/application/FavoriteService.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    public List<Favorite> findFavoritesByMemberId(Long memberId) {
+        return favoriteDao.findByMemberId(memberId);
```
findAllByMemberId
요즘 많이 사용하는 JPA 방식에서는 좀 더 명확하게 findAll, findOne 과 같은 네이밍을 사용합니다.
미세팁으로 그냥 알고만 계시면 좋을 것 같아요
https://docs.spring.io/spring-data/jpa/docs/current/api/org/springframework/data/jpa/repository/JpaRepository.html

htrucci commented on src/main/java/subway/favorite/dao/FavoriteDao.java:
```diff
@@ -0,0 +1,65 @@
+                .usingGeneratedKeyColumns("id");
+    }
+
+    public Favorite insert(Long memberId, Long sourceStationId, Long targetStationId) {
+        Map<String, Object> params = new HashMap<>();
+        params.put("member_id", memberId);
+        params.put("source_station_id", sourceStationId);
+        params.put("target_station_id", targetStationId);
+
+        Long id = simpleJdbcInsert.executeAndReturnKey(params).longValue();
+        return new Favorite(id, memberId, new Station(sourceStationId), new Station(targetStationId));
+    }
```
Map params를 만들지 않고 아래 방식으로 바로 맵핑하여 넣을수도 있습니다 😄 
SqlParameterSource params = new BeanPropertySqlParameterSource(Favorite객체);

sumniy replied:
Member 객체가 sourceStationId, targetStationId가 아니라 sourceStation, targetStation 객체 자체를 가지고 있는 상태라서 Map을 활용해서 넣어주었는데요. 

보통 이렇게 도메인 객체가 DB 테이블 구조와 다를 때는 DB 접근용으로 별도의 Dto를 활용하는게 맞을까요? 
근데 그렇게 하면 Dao에서만 사용하는 Dto 느낌이라 뭔가 과한 느낌이 듭니다..  

이 부분에 대해서 어떻게 생각하시는지 말씀해 주시면 감사하겠습니다.

htrucci replied:
그렇네요, 다시보니 map으로 전달한 이유를 이해했습니다.
사실 Favorite객체의 필드가 테이블과 동일하게 sourceStationId, targetStationId로 갖고있는것이
요즘 추세인 JPA ORM방식에 더 맞다고 볼 수 있고, Dto를 사용했을떄의 관리성 용이도 있겠지요.
그렇다고 Map사용이 잘못됬다는 건 아닙니다. Map의 유연함을 잘 사용해도 좋은데, FAVORITE 테이블에 관계된 데이터만 단순히 Insert하는 부분이어서 가능하면 dto가 아닌 Favorite객체 자체를 이용하는것이 어떨까 싶은 부분이네요.
추가해주신대로 getSourceStationId()를 넣어두면 현재도 SqlParameterSource params = new BeanPropertySqlParameterSource(Favorite객체); 를 적용하는데 문제는 없을 것이고 필요하다면 (enum, 랩핑클래스 사용시) 맵핑을 위해 BeanPropertySqlParameterSource 내부 함수들을 오버로딩 하는 경우도 있을 것입니다.
사실 요즘은 프레임워크가 잘 되어있기때문에 이런 고민을 안해도 되지만,, 어쨋든 Map으로 구현하신 부분도 좋습니다 👍  다만 CRUD가 일어나는 dao마다 map이 생겨야한다면,, DTO가 나을 수도 있겠어요.

sumniy replied:
답변 감사합니다! 말씀해주신대로 Favorite 객체 자체로 BeanPropertySqlParameterSource를 사용하는 방식으로 수정했습니다.

htrucci commented on src/main/java/subway/favorite/dao/FavoriteDao.java:
```diff
@@ -0,0 +1,65 @@
+                new Station((Long) it.get("SOURCE_STATION_ID"), (String) it.get("SOURCE_STATION_NAME")),
+                new Station((Long) it.get("TARGET_STATION_ID"), (String) it.get("TARGET_STATION_NAME"))
+        );
+    }
```
RowMapper를 써볼까요? 👀

htrucci commented on src/main/java/subway/favorite/dao/FavoriteDao.java:
```diff
@@ -0,0 +1,65 @@
+        return new Favorite(id, memberId, new Station(sourceStationId), new Station(targetStationId));
+    }
+
+    public List<Favorite> findByMemberId(Long memberId) {
+        String sql = "select F.id as favorite_id, F.member_id as favorite_member_id, F.source_station_id as favorite_source_station_id, F.target_station_id as favorite_target_station_id, " +
+                "SST.id as source_station_id, SST.name as source_station_name, " +
+                "TST.id as target_station_id, TST.name as target_station_name " +
+                "from FAVORITE F \n" +
+                "join STATION SST on F.source_station_id = SST.id " +
+                "join STATION TST on F.target_station_id = TST.id " +
+                "WHERE F.member_id = ?";
```
저도 쿼리를 짤때 성능이슈가 없다면 가능한 연관 관계를 모두 포함하여 짜곤합니다 👍 
데이터 정합성을 함께 확인할 수 있어서 좋습니다.

htrucci commented on src/main/java/subway/favorite/dto/FavoriteResponse.java:
```diff
@@ -1,11 +1,25 @@
+        this.target = target;
+    }
+
+    public static FavoriteResponse of(Favorite favorite) {
+        return new FavoriteResponse(
+                favorite.getId(),
+                StationResponse.of(favorite.getSourceStation()),
+                StationResponse.of(favorite.getTargetStation()));
+    }
```
💯

htrucci commented on src/main/java/subway/favorite/ui/FavoriteController.java:
```diff
@@ -1,8 +1,48 @@
+        this.stationService = stationService;
+    }
+
+    @PostMapping
+    public ResponseEntity<FavoriteResponse> createFavorite(@AuthenticationPrincipal LoginMember loginMember, @RequestBody FavoriteRequest favoriteRequest) {
+        FavoriteResponse favoriteResponse = FavoriteResponse.of(favoriteService.createFavorite(loginMember.getId(), favoriteRequest.getSource(), favoriteRequest.getTarget()));
+        return ResponseEntity.created(URI.create("/favorites/" + favoriteResponse.getId())).body(favoriteResponse);
+    }
```
서비스 호출 line이 매우 길어졌네요.
createFavorite 서비스에 FavoriteRequest를 통으로 넘기는건 어떤가요? 👀

sumniy replied:
넵 가독성이 좋지 않았네요. 수정했습니다 😀

근데 관련해서 제가 MVC 패턴을 공부하다가 알게된 것에 대해서 질문드리고 싶은게 있습니다.

어떤 블로그에서 Controller에서는 request를 deserialize하고 response를 serialize 해주도록 하면, Service 계층과 Dto 간의 의존성을 분리해서 Service의 재사용성과 책임 분리가 잘 된다고 하는 글을 봤습니다.

개인적으로는 이에 반박할만한 생각이나 경험이 아직까지는 없어서 괜찮은 방법이라고 생각하고 있었는데, 주신 피드백과는 조금은 다른 방향인 것 같아서.. 실무에서는 크게 신경쓰지 않는 부분인지 아니면 어떤 규칙이 있는 것인지, 데빈은 어떻게 생각하시는지 의견 듣고 싶습니다!

htrucci replied:
말씀하신 부분이 favoriteRequest를 deserialize해서 service용 DTO로 변환해주는 내용일까요?  favoriteService.createFavorite(createFavoriteDTO)
아니면 
favoriteService.createFavorite(favoriteRequest)
favoriteService.createFavorite(loginMember.getId(), favoriteRequest.getSource(), favoriteRequest.getTarget()));
위의 차이인가요?
적어주신 내용에 크게 공감을 합니다. Controller레이어와 Service레이어 DTO 분리를 해주면 말씀해주신 장점이 있을 것 같아요.
제가 피드백 드린내용은 이거에 반하는 내용이라기보다 param이 길어지면 안좋아서 DTO로 넘겼으면 좋겠다는 의견이긴 했는데, 사실 DTO분리까지는 생각을 못하였고, Service용 DTO로 변환해주는 과정이 있다면 의존성을 분리하여 재사용성을 늘리지 않을까요?
다만 의존성 분리를 위해.. 아래와 같이 작성되는 부분도 있는데요. 많이 접하는 코드이기도 합니다 -ㅇ-;; 
이런건 지양해야 한다고 생각합니다.
Map<String, Object> favoriteService.createFavorite(Map<String,Object> paramMap)

sumniy replied:
답변 감사합니다! 처음에는 그저 request, response dto만 따로 Service와 오고가지 않는 것을 생각하고 그것에 대해서만 질문을 드렸는데, 제가 너무 단순하게 질문을 드린 것 같네요 ㅎㅎ;; 말씀해주신 Service 용 Dto에 대한 생각까지는 못 했었는데 좋은 의견 말씀해주셔서 감사합니다.

htrucci commented on src/main/java/subway/line/dao/LineDao.java:
```diff
@@ -31,44 +32,50 @@ public Line insert(Line line) {
     public void update(Line newLine) {
         String sql = "update LINE set name = ?, color = ? where id = ?";
-        jdbcTemplate.update(sql, new Object[]{newLine.getName(), newLine.getColor(), newLine.getId()});
+        jdbcTemplate.update(sql, newLine.getName(), newLine.getColor(), newLine.getId());
```
👍

htrucci commented on src/main/java/subway/line/domain/Lines.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class Lines {
+
+    List<Line> lines;
+
+    public Lines(List<Line> lines) {
+        this.lines = lines;
+    }
+
+    public SectionsInAllLine getSectionsInAllLine() {
+        return new SectionsInAllLine(lines.stream()
+                .flatMap(line -> line.getSections()
+                        .getSections()
+                        .stream()
+                        .map(section -> new SectionWithFare(line.getExtraFare(), section)))
+                .collect(Collectors.toList()));
+    }
+}
```
일급컬렉션화를 잘 해주셨네요 💯

htrucci commented on src/main/java/subway/line/domain/SectionsInAllLine.java:
```diff
@@ -0,0 +1,38 @@
+        return sections;
+    }
+
+    private Map<Long, Station> createStationMap() {
+        return sections.stream()
+                .flatMap(section -> section.getStations().stream())
+                .distinct()
+                .collect(Collectors.toMap(Station::getId, Function.identity()));
+    }
```
Function.identity() 는 처음 보아서 저도 찾아보았네요. 
중복을 제거한 이중 for문처럼 돌아가는군요. 잘 배웠습니다. 😄

sumniy replied:
저도 페어분께 배웠습니다 ㅎㅎㅎㅎ 😆

htrucci commented on src/main/java/subway/path/application/PathService.java:
```diff
@@ -0,0 +1,29 @@
+@Service
+public class PathService {
+
+    LineService lineService;
```
private final 을 빼먹으신 듯 하네요 😄

sumniy replied:
체크 감사합니다!

htrucci commented on src/main/java/subway/path/domain/Path.java:
```diff
@@ -0,0 +1,97 @@
+        }
+
+        if (isInfant(age)) {
+            fare = FREE_FARE;
+        }
+
+        return fare;
+    }
+
+    private boolean isTeenager(int age) {
+        return age >= TEENAGER_AGE_LOWERBOUND && age < TEENAGER_AGE_UPPERBOUND;
+    }
+
+    private boolean isChild(int age) {
+        return age >= CHILD_AGE_LOWERBOUND && age < TEENAGER_AGE_LOWERBOUND;
+    }
+
+    private boolean isInfant(int age) {
+        return age < CHILD_AGE_LOWERBOUND;
+    }
+
+    private int getDistanceAdditionalFare() {
+        int distance = getDistance();
+
+        if (distance > SECOND_DISTANCE_SECTION) {
+            return FIRST_DISTANCE_SECTION_ADDITIONAL_FARE +
+                    getAdditionalFare(distance - SECOND_DISTANCE_SECTION, SECOND_DISTANCE_UNIT);
+        }
+
+        if (distance > FIRST_DISTANCE_SECTION) {
+            return getAdditionalFare(distance - FIRST_DISTANCE_SECTION, FIRST_DISTANCE_UNIT);
+        }
+
+        return NO_ADDITIONAL_FARE;
+    }
+
+    private int getAdditionalFare(int distance, double distanceUnit) {
+        return ((int) Math.ceil(distance / distanceUnit)) * ADDITIONAL_FARE_BY_DISTANCE;
+    }
+}
```
거리 체계와 연령별 요금체계 계산이 한 클래스에 모여있어 조금 복잡한 것 같네요.
선언된 상수들도 다소 파악하기 어렵구요 😭 
요금체계가 복잡해지면 손대기 어려운 거대한 Legacy가 될 수도 있겠어요.
PathFare와 AgeFare로 분리해보는 것은 어떨까요?

sumniy replied:
넵 분리해보겠습니다!

htrucci commented on src/main/java/subway/path/domain/SubwayEdge.java:
```diff
@@ -0,0 +1,17 @@
+
+public class SubwayEdge extends DefaultWeightedEdge {
+
+    SubwayWeight weight;
```
protected로 선언하면 문제가 없을까요? 👀

sumniy replied:
라이브러리의 일부만 상속해서 사용하느라 어떤 영향이 갈지 예측하기 힘들어서 기존 라이브러리의 코드 스타일을 따라 작성했습니다. 
이런 경우에는 저희 스타일에 맞게 수정하고 안되는 부분을 찾아 수정하는게 더 나은 방법일까요?

htrucci replied:
아 그렇군요 ㅎㅎ 적어주신 사유라면 잘 고려하신 것 같아요.

htrucci commented on src/main/java/subway/path/exceptions/NotFoundPathException.java:
```diff
@@ -0,0 +1,7 @@
+
+public class NotFoundPathException extends RuntimeException {
+
+    public NotFoundPathException() {
+    }
```
super가 필요하겠어요

sumniy replied:
넵 반영하겠습니다.

htrucci commented on src/main/java/subway/path/exceptions/UnconnectedPathException.java:
```diff
@@ -0,0 +1,12 @@
+
+import subway.station.domain.Station;
+
+public class UnconnectedPathException extends RuntimeException {
+
+    private static final String MESSAGE_FORMAT = "%s에서 %s으로 가는 경로를 찾을 수 없습니다.";
+
+    public UnconnectedPathException(Station source, Station target) {
+        super(String.format(MESSAGE_FORMAT, source.getName(), target.getName()));
+    }
+}
```
👍

htrucci commented on src/main/java/subway/path/ui/PathController.java:
```diff
@@ -1,5 +1,50 @@
+                                                         @AuthenticationPrincipal(required = false) LoginMember loginMember) {
+        Path shortestPath = pathService.getShortestPath(source, target);
+
+        PathResponse pathResponse = new PathResponse(
+                shortestPath.getStations()
+                        .stream()
+                        .map(StationResponse::of)
+                        .collect(Collectors.toList()), shortestPath.getDistance(),
+                shortestPath.getFare(loginMember.getAge()));
```
shortestPath.getStations()
                        .stream()
                        .map(StationResponse::of)
                        .collect(Collectors.toList()
위 코드를 PathResponse의 생성자 로직으로 옮기면 컨트롤러가 더 깔끔해질 것 같아요.
그리고 shortestPath.getFare() 를 호출하거나 하는 부분은 service레이어로 들어가는 것이 좋을 것 같아요.

sumniy replied:
넵 반영하겠습니다

htrucci commented on src/main/java/subway/path/ui/PathController.java:
```diff
@@ -1,5 +1,50 @@
+
+    PathService pathService;
+
+    @Autowired
+    public PathController(PathService pathService) {
+        this.pathService = pathService;
+    }
```
스프링 4.3부터 단일 생성자일 경우는 @Autowired를 붙이지 않아도 됩니다.
저는 습관적으로 그냥 다 붙여주긴 합니다 👀

sumniy replied:
아하 ㅎㅎ 저도 위에서 두개 이상일 때 붙이는 가이드 알려주신거 반영하다가 다 붙이는게 일관성도 있어보이고 한눈에 보이기도 해서 다 붙였습니다 ㅎㅎ

htrucci commented on src/main/java/subway/path/ui/PathController.java:
```diff
@@ -1,5 +1,50 @@
+        return ResponseEntity.ok(pathResponse);
+    }
+
+    @ExceptionHandler(Exception.class)
+    public ResponseEntity exceptionHandle(Exception e) {
+        e.printStackTrace();
+
+        return ResponseEntity.badRequest().body(e.getMessage());
+    }
```
앗, SubwayAdvice로 옮겨주세요 😄 
모든 Exception에 대한 처리면 400보다 500에러가 적합하지 않을까요?
@ApiResponse를 사용하는 부분도 고려해주세요 👍

sumniy replied:
넵! 고려해보고 같이 반영해보겠습니다.

htrucci reviewed (changes requested):
안녕하세요, 터너! 아주 반가운 이름이네요😄😄
같이 지하철 서비스 미션 리뷰 하게된 데빈입니다. 
모든 기능이 정상적으로 동작하고 아주 잘 구현해주셨습니다 👍 👍 
리뷰 남겨드린 내용 한번 고민해보시고 다시 PR 부탁드립니다~! 😄

sumniy committed:
Refactor: 피드백 반영 - @Autowired

스프링 재단 가이드에 따라 두개 이상의 객체가 주입될 때
@Autowired 어노테이션을 붙이도록 변경.
추가로 일관성, 가독성을 위해 모든 객체 주입 생성자에
@Autowired 어노테이션을 붙힘.

sumniy committed:
Refactor: 피드백 반영 - 패키지 구조, ExceptionHandler

Application 파일을 제외한 파일들을 다른 패키지로 이동.
ExceptionHandler에서 @ResponseStatus 어노테이션을 활용하여
간단하게 response code를 반환하도록 변경.

sumniy committed:
Refactor: 피드백 반영 - createFavorite의 station 유효성 검사 추가

createFavorite에서 favorite를 생성하기 전에
stationId들을 가지고 station에 대한 유효성 검사를 하도록 변경.

sumniy committed:
Refactor: 피드백 반영 - favoriteDao find에서 rowMapper 적용

rowMapper를 적용하여, 불필요한 stream과 코드를 줄임.

sumniy committed:
Refactor: 피드백 반영 - style, 접근 제한자 수

sumniy committed:
Refactor: 피드백 반영 - Path의 거리별, 나이별 요금 계산 분리

거리별 추가요금 계산 로직과, 나이대별 요금 할인 로직을
각각의 enum 클래스로 분리하여 상수값들을 의미있게 묶고
책임을 분리함.

sumniy committed:
Refactor: 피드백 반영 - Path Controller 관련

PathController의 ExceptionHandler를 SubwayExceptionHandler로 이동.
그래프 라이브러리 관련 클래스 libs 패키지로 관리.
controller에서 path객체의 메서드를 호출하지 않고 Service를 거치도록 수정.
PathResponse 생성 관련 로직 일부를 Controller에서 PathResponse 내부로 이동.

sumniy committed:
Refactor: 피드백 반영 - 로그인 로직 개선

email만으로 가져오는것이 아닌 email과 password로
member를 찾아서 password 유출의 위험을 방지.

sumniy commented:
안녕하세요 데빈, 터너입니다.

꼼꼼하게 리뷰해주셔서 감사합니다. 

작성해주신 피드백 반영했습니다. 그리고 몇몇 피드백에 질문 남겼는데 답변 해주시면 감사하겠습니다. 🙇‍♂️

htrucci commented on src/main/java/subway/member/ui/MemberController.java:
```diff
@@ -42,19 +47,22 @@ public ResponseEntity createMember(@RequestBody MemberRequest request) {
     }
```
미로그인 유저도 회원seq만 유추한다면 계정정보를 보거나 마음대로 삭제가 가능하겠어요.
자신의 계정만 삭제가 가능하도록 보안처리가 필요하겠네요 😄

sumniy replied:
앗 정말이네요. 따로 수정하겠습니다!

htrucci reviewed (approved):
안녕하세요 터너, 데빈입니다.
리팩토링 된 부분 잘 보았습니다 👍 
특별히 추가 코멘트는 없을 것 같아 머지하도록 하겠습니다.
고생하셨습니다~! 👏

sumniy commented:
꼼꼼한 리뷰 감사합니다~
